include/eos/fitting/fitting.hpp

```diff
@@ -778,6 +778,155 @@ inline std::pair<core::Mesh, fitting::RenderingParameters> fit_shape_and_pose(
+ * @param[in] lambda_identity Regularisation parameter of the PCA shape fitting.
+ * @param[in] num_expression_coefficients_to_fit How many shape-coefficients to fit (all others will stay 0). Should be bigger than zero, or std::nullopt to fit all coefficients. Only used for expression-PCA fitting.
+ * @param[in] lambda_expressions Regularisation parameter of the expression fitting. Only used for expression-PCA fitting.
+ * @param[in,out] pca_shape_coefficients If given, will be used as initial PCA shape coefficients to start the fitting. Will contain the final estimated coefficients.
+ * @param[in,out] expression_coefficients If given, will be used as initial expression blendshape coefficients to start the fitting. Will contain the final estimated coefficients.
+ * @param[out] fitted_image_points Debug parameter: Returns all the 2D points that have been used for the fitting.
+ * @return The fitted model shape instance and the final pose.
+ */
+inline std::pair<core::Mesh, fitting::RenderingParameters> fit_shape_and_pose(
+    const morphablemodel::MorphableModel& morphable_model,
+    const core::LandmarkCollection<Eigen::Vector2f>& landmarks, const core::LandmarkMapper& landmark_mapper,
+    int image_width, int image_height, int num_iterations,
+    cpp17::optional<int> num_shape_coefficients_to_fit, float lambda_identity,
+    cpp17::optional<int> num_expression_coefficients_to_fit, cpp17::optional<float> lambda_expressions,
+    std::vector<float>& pca_shape_coefficients, std::vector<float>& expression_coefficients,
+    std::vector<Eigen::Vector2f>& fitted_image_points)
+{
+    assert(landmarks.size() >= 4);
+    assert(image_width > 0 && image_height > 0);
+    assert(num_iterations > 0); // Can we allow 0, for only the initial pose-fit?
+    assert(pca_shape_coefficients.size() <= morphable_model.get_shape_model().get_num_principal_components());
+    // More asserts I forgot?
+    if (num_shape_coefficients_to_fit)
+    {
+        if (num_shape_coefficients_to_fit.value() >
+            morphable_model.get_shape_model().get_num_principal_components())
+        {
+            throw std::runtime_error(
+                "Specified more shape coefficients to fit than the given shape model contains.");
+        }
+    }
+
+    using Eigen::MatrixXf;
+    using Eigen::Vector2f;
+    using Eigen::Vector4f;
+    using Eigen::VectorXf;
+    using std::vector;
+
+    // The 2D and 3D point correspondences used for the fitting:
+    vector<int> vertex_indices; // their vertex indices
+    vector<Vector2f> image_points; // the corresponding 2D landmark points
+
+    // Sub-select all the landmarks which we have a mapping for (i.e. that are defined in the 3DMM),
+    // and get the corresponding model points (mean if given no initial coeffs, from the computed shape otherwise):
+    for (int i = 0; i < landmarks.size(); ++i)
```

Comment:
As discussed, I've removed the duplicated fitting loop leaving only the landmark mapper loop. I've also removed a few lines that get the model points, current pca shape and current combined shape because all of these operations already take place in the existing function called at line 891.